third_party/tlx/dialect/lib/Transforms/InsertRequireLayout.cpp

```diff
@@ -0,0 +1,99 @@
+#include "IR/Dialect.h"
+#include "mlir/Analysis/SliceAnalysis.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "triton/Dialect/Triton/IR/Dialect.h"
+#include "triton/Dialect/TritonGPU/IR/Dialect.h"
+#include "triton/Dialect/TritonGPU/IR/Types.h"
+#include "triton/Dialect/TritonGPU/Transforms/Passes.h"
+#include "triton/Dialect/TritonGPU/Transforms/Utility.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Transforms/Passes.h"
+
+
+#define DEBUG_TYPE "tlx-amd-insert-require-layout"
+#define DBGS() (llvm::dbgs() << "[" DEBUG_TYPE "]: ")
+#define LDBG(X) LLVM_DEBUG(DBGS() << X << "\n")
+
+using namespace mlir;
+namespace tt = ::mlir::triton;
+namespace ttg = ::mlir::triton::gpu;
+namespace tlx = ::mlir::triton::tlx;
+
+namespace mlir {
+
+std::optional<ttg::SwizzledSharedEncodingAttr>
+getSharedEncIfAllUsersAreDotEnc(Value loadedValue);
+
+namespace triton {
+namespace tlx {
+
+#define GEN_PASS_DEF_TLXINSERTREQUIRELAYOUT
+#include "tlx/dialect/include/Transforms/Passes.h.inc"
+
+LogicalResult insertRequireLayout(ModuleOp m) {
+  OpBuilder builder(m.getContext());
+  LDBG("insertRequiredLayout\n");
+  WalkResult result = m.walk([&](tt::DotOp dotOp) -> WalkResult {
+    SetVector<Operation *> backwardSet;
+    BackwardSliceOptions options;
+    options.inclusive = false;
+    options.omitUsesFromAbove = false;
+    mlir::getBackwardSlice(dotOp.getOperation(), &backwardSet, options);
+    LLVM_DEBUG({
+          llvm::dbgs() << "DotOp\n";
+          dotOp.dump();
+    });
+    for (Operation *op : backwardSet) {
+      if (auto localLoadOp = dyn_cast<ttg::LocalLoadOp>(op)) {
+        LLVM_DEBUG({
+          llvm::dbgs() << "LocalLoadOp\n";
+          localLoadOp.dump();
+        });
+        // Get the shared encoding for this local load op based on the dot op
+        auto encoding = getSharedEncIfAllUsersAreDotEnc(localLoadOp->getResult(0)).value_or(nullptr);
```

Comment:
A link dependency might be needed due to this call, as I'm hitting

third_party/tlx/dialect/lib/Transforms/InsertRequireLayout.cpp:54: undefined reference to `mlir::getSharedEncIfAllUsersAreDotEnc(mlir::Value)'